reciprocalspaceship/io/crystfel.py

```diff
@@ -1,270 +1,553 @@
+
+    @property
+    def available_chunk_metadata_keys(self) -> list:
+        """Keys which can be passed to parallel_read_crystfel to customize the chunk level metadata"""
+        return list(self.re_chunk_metadata.keys())
+
+    @property
+    def available_crystal_metadata_keys(self) -> list:
+        """Keys which can be passed to parallel_read_crystfel to customize the crystal level metadata"""
+        return list(self.re_crystal_metadata.keys())
+
+    def parallel_read_crystfel(
+        self,
+        wavelength=None,
+        chunk_metadata_keys=None,
+        crystal_metadata_keys=None,
+        peak_list_columns=None,
+        **ray_kwargs,
+    ) -> list:
+        """
+        Parse a CrystFEL stream file using multiple processors. Parallelization depends on the ray library (https://www.ray.io/).
+        If ray is unavailable, this method falls back to serial processing on one CPU. Ray is not a dependency of reciprocalspaceship
+        and will not be installed automatically. Users must manually install it prior to calling this method.
+
+        PARAMETERS
+        ----------
+        wavelength : float
+            Override the wavelength with this value. Wavelength is used to compute Ewald offsets.
+        chunk_metadata_keys : list
+            A list of metadata_keys which will be returned in the resulting dictionaries under the 'chunk_metadata' entry.
+            A list of possible keys is stored as stream_loader.available_chunk_metadata_keys
+        crytal_metadata_keys : list
+            A list of metadata_keys which will be returned in the resulting dictionaries under the 'crystal_metadata' entry.
+            A list of possible keys is stored as stream_loader.available_crystal_metadata_keys
+        peak_list_columns : list
+            A list of columns to include in the peak list numpy arrays.
+            A list of possible column names is stored as stream_loader.available_column_names.
+        ray_kwargs : optional
+            Additional keyword arguments to pass to [ray.init](https://docs.ray.io/en/latest/ray-core/api/doc/ray.init.html#ray.init).
+
+        RETURNS
+        -------
+        chunks : list
+            A list of dictionaries containing the per-chunk data. The 'peak_lists' item contains a
+            numpy array with shape n x 14 with the following information.
+                h, k, l, I, SIGI, peak, background, fs/px, ss/px, s1x, s1y, s1z,
+                ewald_offset, angular_ewald_offset
+        """
+        if peak_list_columns is not None:
+            peak_list_columns = [self.peak_list_columns[s] for s in peak_list_columns]
+
+        with open(self.filename, "r+") as f:
+            memfile = mmap(f.fileno(), 0)
+            beginnings_and_ends = zip(
+                self.block_regex_bytes["chunk_begin"].finditer(memfile),
+                self.block_regex_bytes["chunk_end"].finditer(memfile),
+            )
+            try:
+                import ray
```

Comment:
i added an option to disable ray, and i changed the code to check before trying to import ray (84b83ad07b9024ccc3bf8f42f46ad3dfda2ef35c). if we end up with more functions that rely on ray, we can set the environment variable on module import. however, i want to leave this as is right now in case someone decides to do something like `!pip install ray` from an active IPython terminal or Jupyter notebook.